icepyx/core/types.py

```diff
@@ -0,0 +1,111 @@
+from __future__ import annotations
+
+from typing import Literal, TypedDict, Union
+
+from typing_extensions import NotRequired
+
+IcesatProductShortName = Literal[
+    "ATL01",
```

Comment:
Got any suggested resources or keywords for reading up on type-checking versus input validation? I don't have the vocabulary to google beyond basics, and while I understand it's useful to check more than "is it a string?", to me this sort of type declaration should effectively be replacing the `is2ref._validate_product()` function (or better yet, turning that function into this plus a helpful error message - can you add custom messages to type checking?) rather than maintaining the list of literals in multiple spots.

Reply:
>Got any suggested resources or keywords for reading up on type-checking versus input validation?

I don't have a resource off the top of my head, but here's my quick explanation, and I'm always happy to chat about type checking :)

Type checking occurs without running the code. A type checker is like you or me reading the code and saying "That doesn't look right, isn't that supposed to be a string?" Except we tell the type checker explicitly "This is supposed to be a string" and let it read the code for us to look for instances where we're doing something we're not supposed to do with that thing we told the typechecker is supposed to be a string. Like using the division operator on it! This type will prevent us from writing code that uses an invalid short name, but won't catch users passing in an invalid short name at runtime.

Input validation will catch the latter case, but of course requires dilligence on our part to write the input validations and maintain them. We can get the best of both worlds with awesome tools like [pydantic's validation decorator](https://docs.pydantic.dev/latest/concepts/validation_decorator/) (beartype and other libraries can do this too!). Because type annotations add metadata to Python objects, Pydantic can inspect that metadata at runtime and generate input validation code for us! :exploding_head: 

>to me this sort of type declaration should effectively be replacing the is2ref._validate_product() function (or better yet, turning that function into this plus a helpful error message ...

:100: I think this is a wonderful use case for `@validate_call` or similar :) 

> ... can you add custom messages to type checking?

The pydantic decorator shown above doesn't seem to allow customizing the error messages it generates from runtime validations, but the error messages are very helpful in my experience. But please note that this different than type checking, it's runtime input validation based on type annotation metadata :)

Reply:
Thanks for this explanation. So it sounds like we can only maintain the list in one spot, it just needs to live on the type-checking side (plus a decorator for validating) rather than the validation side (plus a decorator for type checking). Looking forward to making this simplification in a future PR!